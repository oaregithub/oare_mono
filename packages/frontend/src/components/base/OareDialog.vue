<template>
  <div>
    <v-dialog
      :value="value"
      @input="$emit('input', $event)"
      width="500"
      persistent
    >
      <template v-slot:activator="{ on }">
        <slot name="activator" :on="on"></slot>
      </template>

      <v-card>
        <v-card-title data-testid="dialog-title">{{ title }}</v-card-title>
        <v-card-text>
          <slot></slot>
        </v-card-text>
        <v-divider />
        <v-card-actions>
          <v-spacer />
          <v-btn
            @click="$emit('input', false)"
            color="error"
            text
            data-testid="cancel-btn"
            >{{ cancelText }}</v-btn
          >

          <slot name="submit-button" :on="{ click: submit }">
            <OareLoaderButton
              @click="submit"
              color="primary"
              :loading="submitDisabled"
              data-testid="submit-btn"
            >
              {{ submitText }}
            </OareLoaderButton>
          </slot>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "@vue/composition-api";

export default defineComponent({
  name: "OareDialog",
  props: {
    value: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      required: true,
    },
    submitText: {
      type: String,
      default: "Submit",
    },
    cancelText: {
      type: String,
      default: "Cancel",
    },
    submitLoading: {
      type: Boolean,
      default: false,
    },
    submitDisabled: {
      type: Boolean,
      default: false,
    },
    closeOnSubmit: {
      type: Boolean,
      default: false,
    },
  },
  setup({ closeOnSubmit }, { emit }) {
    function submit() {
      if (closeOnSubmit) {
        emit("input", false);
      }
      emit("submit");
    }

    return {
      submit,
    };
  },
});
</script>
