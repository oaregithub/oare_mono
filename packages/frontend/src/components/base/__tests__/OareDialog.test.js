import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import OareDialog from '../OareDialog.vue';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('OareDialog', () => {
  const createWrapper = (props = {}) =>
    mount(OareDialog, {
      localVue,
      vuetify,
      propsData: {
        title: 'Test Dialog',
        value: true,
        ...props,
      },
    });

  test('matches snapshot', () => {
    const wrapper = createWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  test('shows title', () => {
    const wrapper = createWrapper();
    const title = wrapper.find('[data-testid="dialog-title"]');
    expect(title.text()).toBe('Test Dialog');
  });

  test('default submit text is "Submit"', () => {
    const wrapper = createWrapper();
    const submitBtn = wrapper.find('.test-submit-btn');
    expect(submitBtn.text()).toBe('Submit');
  });

  test('default cancel text is "Cancel"', () => {
    const wrapper = createWrapper();
    const cancelBtn = wrapper.find('[data-testid="cancel-btn"');
    expect(cancelBtn.text()).toBe('Cancel');
  });

  test('override default submit text', () => {
    const wrapper = createWrapper({
      submitText: 'Custom submit',
    });
    const submitBtn = wrapper.find('.test-submit-btn');
    expect(submitBtn.text()).toBe('Custom submit');
  });

  test('override default cancel text', () => {
    const wrapper = createWrapper({
      cancelText: 'Custom cancel',
    });
    const cancelBtn = wrapper.find('[data-testid="cancel-btn"]');
    expect(cancelBtn.text()).toBe('Custom cancel');
  });

  test('submitLoading disables submit button', () => {
    const wrapper = createWrapper({
      submitLoading: true,
    });
    const submitBtn = wrapper.find('.test-submit-btn').element;
    expect(submitBtn).toBeDisabled();
  });
});
