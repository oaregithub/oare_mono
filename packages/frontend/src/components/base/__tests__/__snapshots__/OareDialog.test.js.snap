// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OareDialog matches snapshot 1`] = `
<div>
  <v-dialog-stub opendelay="0" closedelay="0" value="true" contentclass="" maxwidth="none" origin="center center" persistent="true" retainfocus="true" transition="dialog-transition" width="500">
    <v-card-stub loaderheight="4" tag="div">
      <v-card-title-stub datatestid="dialog-title" data-testid="dialog-title">Test Dialog</v-card-title-stub>
      <v-card-text-stub></v-card-text-stub>
      <v-divider-stub></v-divider-stub>
      <v-card-actions-stub>
        <v-spacer-stub></v-spacer-stub>
        <v-btn-stub color="error" tag="button" activeclass="" text="true" type="button" data-testid="cancel-btn">Cancel</v-btn-stub>
        <oareloaderbutton-stub color="primary" data-testid="submit-btn">
          Submit
        </oareloaderbutton-stub>
      </v-card-actions-stub>
    </v-card-stub>
  </v-dialog-stub>
</div>
`;
